zerver/views/upload.py

```diff
@@ -25,14 +26,37 @@
 from zerver.models import UserProfile, validate_attachment_request
 
 
+def patch_disposition_header(response: HttpResponse, url: str, is_attachment: bool) -> None:
+    disposition = "attachment" if is_attachment else "inline"
+
+    filename = os.path.basename(urlparse(url).path)
+
+    try:
+        filename.encode("ascii")
+        file_expr = 'filename="{}"'.format(filename.replace("\\", "\\\\").replace('"', r"\""))
+    except UnicodeEncodeError:
+        file_expr = "filename*=utf-8''{}".format(quote(filename))
+
+    response.headers["Content-Disposition"] = f"{disposition}; {file_expr}"
+
+
 def serve_s3(
     request: HttpRequest, url_path: str, url_only: bool, download: bool = False
 ) -> HttpResponse:
-    url = get_signed_upload_url(url_path, download=download)
+    mimetype, encoding = guess_type(url_path)
+    attachment = download or mimetype not in INLINE_MIME_TYPES
+
+    url = get_signed_upload_url(url_path, download=attachment)
+    assert url.startswith("https://")
     if url_only:
-        return json_success(request, data=dict(url=url))
+        return json_success(request, data=dict(url="XXX FIXME"))
```

Comment:
I think we need to use `generate_unauthed_file_access_url` here like `serve_local` does, and then generalize `serve_local_file_unauthed` not to be local-only?

Or we could send a signed S3 URL as before, as long as we remember to keep it short-lived even if we use a longer-lived URL for our internal proxy.